packages/backend/src/modules/finality/analyzers/ScrollFinalityAnalyzer.ts

```diff
@@ -0,0 +1,98 @@
+import { assert, TrackedTxsConfigSubtype, UnixTime } from '@l2beat/shared-pure'
+import { utils } from 'ethers'
+import { z } from 'zod'
+
+import { BaseAnalyzer } from './types/BaseAnalyzer'
+
+const ScrollBatchCommit = z.object({
+  version: z.number(),
+  parentBatchHeader: z.string(),
+  chunks: z.array(z.string()),
+  skippedL1MessageBitmap: z.string(),
+})
+type ScrollBatchCommit = z.infer<typeof ScrollBatchCommit>
+
+export class ScrollFinalityAnalyzer extends BaseAnalyzer {
+  override getTrackedTxSubtype(): TrackedTxsConfigSubtype {
+    return 'batchSubmissions'
+  }
+
+  async getFinality(transaction: {
+    txHash: string
+    timestamp: UnixTime
+  }): Promise<number[]> {
+    assert(this.l2Provider, 'Scroll RPC provider not defined')
+
+    const tx = await this.provider.getTransaction(transaction.txHash)
+    const l1Timestamp = transaction.timestamp
+
+    const decodedTransactionData = this.decodeTransaction(tx.data)
+
+    const blockNumbers = decodedTransactionData.chunks
+      .map((chunk) => this.decodeChunkBlockNumbers(chunk))
+      .flat()
+
+    const firstBlockInData = Math.max(...blockNumbers)
+    const lastBlockInData = Math.min(...blockNumbers)
+
+    const l2Blocks = await Promise.all([
```

Comment:
No need to call l2Rpc here. You can decode the timestamp from the block comtext in the batch: https://github.com/scroll-tech/scroll/blob/develop/contracts/src/libraries/codec/ChunkCodecV1.sol#L22-L23

Reply:
Also, this does not change across the versions: https://github.com/scroll-tech/scroll/blob/develop/contracts/src/libraries/codec/ChunkCodecV0.sol#L20-L21

Reply:
L2 calls dropped